instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/TripSessionsBillingTest.kt

```diff
@@ -0,0 +1,311 @@
+
+        mapboxNavigation.startTripSessionAndWaitForFreeDriveState()
+        mapboxNavigation.stopTripSessionAndWaitForIdleState()
+
+        mapboxNavigation.onDestroy()
+
+        delay(HTTP_EVENT_APPEARANCE_TIMEOUT_MILLIS)
+
+        val billingRequests = observer.billingRequests()
+        assertEquals(1, billingRequests.size)
+        assertTrue(billingRequests.first().isFreeDriveSession)
+
+        assertSessionDuration(0, billingRequests.first().duration)
+    }
+
+    @Test
+    fun testBillingSessionsForMixedFreeDriveAndActiveGuidanceTrip() = sdkTest {
+        val observer = HttpServiceEventsObserver.register()
+
+        mapboxNavigation.startTripSessionAndWaitForFreeDriveState()
+        delay(TimeUnit.SECONDS.toMillis(SESSION_DURATION_SECONDS))
+        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(listOf(route))
+        delay(TimeUnit.SECONDS.toMillis(SESSION_DURATION_SECONDS))
+        mapboxNavigation.clearNavigationRoutesAndWaitForUpdate()
+        delay(TimeUnit.SECONDS.toMillis(SESSION_DURATION_SECONDS))
+        mapboxNavigation.onDestroy()
+
+        delay(HTTP_EVENT_APPEARANCE_TIMEOUT_MILLIS)
+
+        val billingRequests = observer.billingRequests()
+        assertEquals(3, billingRequests.size)
+        assertTrue(billingRequests[0].isFreeDriveSession)
+        assertSessionDuration(SESSION_DURATION_SECONDS, billingRequests[0].duration)
+        assertTrue(billingRequests[1].isActiveGuidanceSession)
+        assertSessionDuration(SESSION_DURATION_SECONDS, billingRequests[1].duration)
+        assertTrue(billingRequests[2].isFreeDriveSession)
+        assertSessionDuration(SESSION_DURATION_SECONDS, billingRequests[2].duration)
+    }
+
+    @Test
+    fun testBillingSessionsForStoppedAndResumedActiveGuidanceTrip() = sdkTest {
+        val observer = HttpServiceEventsObserver.register()
+
+        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(listOf(route))
+        mapboxNavigation.startTripSessionAndWaitForActiveGuidanceState()
+        mapboxNavigation.clearNavigationRoutesAndWaitForUpdate()
+
+        delay(TimeUnit.SECONDS.toMillis(SESSION_PAUSE_DURATION_SECONDS))
+
+        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(listOf(route))
+        mapboxNavigation.startTripSessionAndWaitForActiveGuidanceState()
+        mapboxNavigation.clearNavigationRoutesAndWaitForUpdate()
+
+        delay(HTTP_EVENT_APPEARANCE_TIMEOUT_MILLIS)
+
+        val billingRequests = observer.billingRequests()
+        assertEquals(3, billingRequests.size)
+
+        assertTrue(billingRequests[0].isActiveGuidanceSession)
+        assertSessionDuration(0, billingRequests[0].duration)
+        assertTrue(billingRequests[1].isFreeDriveSession)
+        assertSessionDuration(SESSION_PAUSE_DURATION_SECONDS, billingRequests[1].duration)
+        assertTrue(billingRequests[2].isActiveGuidanceSession)
+        assertSessionDuration(0, billingRequests[2].duration)
+    }
```

Comment:
As I see the test name says that it checks pausing and resuming of active guidance session. But there were 2 active guidance sessions in the results. I updated the test so that it verifies pause and resume of a single active guidance. Does it make sence? Or the error is in the test name? 🙂 
```suggestion
    fun testBillingSessionsForStoppedAndResumedActiveGuidanceTrip() = sdkTest {
        val observer = HttpServiceEventsObserver.register()

        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(listOf(route))
        mapboxNavigation.startTripSessionAndWaitForActiveGuidanceState()
        delay(500)
        mapboxNavigation.stopTripSessionAndWaitForIdleState()

        delay(2000)

        mapboxNavigation.startTripSessionAndWaitForActiveGuidanceState()
        delay(500)
        mapboxNavigation.clearNavigationRoutesAndWaitForUpdate()

        delay(HTTP_EVENT_APPEARANCE_TIMEOUT_MILLIS)

        val billingRequests = observer.billingRequests()
        assertEquals(1, billingRequests.size)

        assertTrue(billingRequests[0].isActiveGuidanceSession)
        assertSessionDuration(1, billingRequests[0].duration)
    }
```

Reply:
We already have such a test: `testOneBillingSessionSentForOneSecondLongPausedAndResumedActiveGuidanceTrip`.
But I agree that the name is a bit off. I'm also not sure what we are trying to check here. We have a test that checks paused AG (`testOneBillingSessionSentForOneSecondLongPausedAndResumedActiveGuidanceTrip`) and we have a test that checks there are 3 sessions (`testBillingSessionsForMixedFreeDriveAndActiveGuidanceTrip`).

Reply:
Agree, this test is not necessary. I added it because of my current lack of knowledge about session logic :) I was surprised that we have 3 sessions at the end.